src/linker/Linker.Dataflow/DynamicallyAccessedMembersTypeHierarchy.cs

```diff
@@ -0,0 +1,197 @@
+		readonly Dictionary<TypeDefinition, (DynamicallyAccessedMemberTypes annotation, bool applied)> _typesInDynamicallyAccessedMembersHierarchy;
+
+		public DynamicallyAccessedMembersTypeHierarchy (LinkContext context, MarkStep markStep)
+		{
+			_context = context;
+			_markStep = markStep;
+			_typesInDynamicallyAccessedMembersHierarchy = new Dictionary<TypeDefinition, (DynamicallyAccessedMemberTypes, bool)> ();
+		}
+
+		public (DynamicallyAccessedMemberTypes annotation, bool applied) ProcessMarkedTypeForDynamicallyAccessedMembersHierarchy (TypeDefinition type)
+		{
+			// Non-interfaces must be marked already
+			Debug.Assert (type.IsInterface || _context.Annotations.IsMarked (type));
+
+			DynamicallyAccessedMemberTypes annotation = _context.Annotations.FlowAnnotations.GetTypeAnnotation (type);
+			bool apply = false;
+
+			// We'll use the cache also as a way to detect and avoid recursion
+			// There's no possiblity to have recursion among base types, so only do this for interfaces
+			if (type.IsInterface) {
+				if (_typesInDynamicallyAccessedMembersHierarchy.TryGetValue (type, out var existingValue))
+					return existingValue;
+
+				_typesInDynamicallyAccessedMembersHierarchy.Add (type, (annotation, false));
+			}
+
+			// Base should already be marked (since we're marking its derived type now)
+			// so we should already have its cached values filled.
+			TypeDefinition baseType = type.BaseType?.Resolve ();
+			Debug.Assert (baseType == null || _context.Annotations.IsMarked (baseType));
+			if (baseType != null && _typesInDynamicallyAccessedMembersHierarchy.TryGetValue (baseType, out var baseValue)) {
+				annotation |= baseValue.annotation;
+				apply |= baseValue.applied;
+			}
+
+			// For the purposes of the DynamicallyAccessedMembers type hierarchies
+			// we consider interfaces of marked types to be also "marked" in that 
+			// their annotations will be applied to the type regardless if later on
+			// we decide to remove the interface. This is to keep the complexity of the implementation
+			// relatively low. In the future it could be possibly optimized.
+			if (type.HasInterfaces) {
+				foreach (InterfaceImplementation iface in type.Interfaces) {
+					var interfaceType = iface.InterfaceType.Resolve ();
+					if (interfaceType != null) {
+						var interfaceValue = ProcessMarkedTypeForDynamicallyAccessedMembersHierarchy (interfaceType);
+						annotation |= interfaceValue.annotation;
+						apply |= interfaceValue.applied;
+					}
+				}
+			}
+
+			Debug.Assert (!apply || annotation != DynamicallyAccessedMemberTypes.None);
+
+			if (apply) {
+				// One of the base/interface types is already marked as having the annotation applied
+				// so we need to apply the annotation to this type as well
+				var reflectionMethodBodyScanner = new ReflectionMethodBodyScanner (_context, _markStep);
+				var reflectionPatternContext = new ReflectionPatternContext (_context, true, type, type);
```

Comment:
```suggestion
				using var reflectionPatternContext = new ReflectionPatternContext (_context, true, type, type);
```

Reply:
It's actually not possible to use `using` since we're passing it as `ref` (and the compiler checks that the values in `using` is immutable).
But it's a valid concern that we're not calling `Dispose` even though it's just a debug check. I modified all call sites to call `Dispose` - without the try/finally as it doesn't seem necessary - But I can be convinced to add that as well.